python/lsst/daf/butler/registry/queries/expressions.py

```diff
@@ -22,19 +22,78 @@
 
 __all__ = ()  # all symbols intentionally private; for internal package use.
 
-from typing import Any, List, Optional, Tuple, TYPE_CHECKING, Union
+from typing import Any, List, Mapping, Optional, Tuple, TYPE_CHECKING, Union
 
 import sqlalchemy
+from sqlalchemy.ext.compiler import compiles
 
 from ...core import DimensionUniverse, Dimension, DimensionElement, NamedKeyDict, NamedValueSet
+from ...core.ddl import AstropyTimeNsecTai
 from .exprParser import Node, TreeVisitor
 from ._structs import QueryColumns
 
 if TYPE_CHECKING:
     import astropy.time
 
 
-def categorizeIdentifier(universe: DimensionUniverse, name: str) -> Tuple[DimensionElement, Optional[str]]:
+class _TimestampColumnElement(sqlalchemy.sql.ColumnElement):
+    """Special ColumnElement type used for TIMESTAMP columns in expressions.
+
+    TIMESTAMP columns in expressions are usually compared to time literals
+    which are `astropy.time.Time` instances that are converted to integer
+    nanoseconds since Epoch. For comparison we need to convert TIMESTAMP
+    column value to the same type. This type is a wrapper for actual column
+    that has special dialect-specific compilation methods defined below
+    transforming column in that common type.
+
+    This mechanism is only used for expressions in WHERE clause, values of the
+    TIMESTAMP columns returned from queries are still handled by standard
+    mechanism and they are converted to `datetime` instances.
+    """
+    def __init__(self, column: sqlalchemy.sql.ColumnElement):
+        super().__init__()
+        self._column = column
+
+
+@compiles(_TimestampColumnElement, "sqlite")
+def compile_timestamp_sqlite(element: Any, compiler: Any, **kw: Mapping[str, Any]) -> str:
+    """Compilation of TIMESTAMP column for SQLite.
+
+    SQLite defines ``strftime`` function that can be used to convert timestamp
+    value to Unix seconds.
+    """
+    return f"STRFTIME('%s', {element._column.name})*1000000000"
+
+
+@compiles(_TimestampColumnElement, "postgresql")
+def compile_timestamp_pg(element: Any, compiler: Any, **kw: Mapping[str, Any]) -> str:
+    """Compilation of TIMESTAMP column for PostgreSQL.
+
+    PostgreSQL can use `EXTRACT(epoch FROM timestamp)` function.
+    """
+    return f"EXTRACT(epoch FROM {element._column.name})*1000000000"
+
+
+def categorizeIngestDateId(name: str) -> bool:
+    """Categorize an identifier in a parsed expression as an ingest_date
+    attribute of a dataset table.
+
+    Parameters
+    ----------
+    name : `str`
+        Identifier to categorize.
+
+    Returns
+    -------
+    isIngestDate : `bool`
+        True is returned if identifier name is ``ingest_date``.
+    """
+    # TODO: this is hardcoded for now, may be better to extract it from schema
+    # but I do not know how to do it yet.
```

Comment:
I wouldn't worry about this too much; I generally haven't so far - hard-coded strings aren't great, but compared to the opaque levels of indirection involved in avoiding them when referring to column names, I think they're not really that bad bad.